Allow to read density without matching hash

sofiasanz opened the pull request:
I find it a little inconvenient that the current `read_density` function only reads the density if there is matching hash. With this PR I now allow for reading the density from `netCDF` files even in cases where the hash doesn't match.

sofiasanz commented on hubbard/ncsile.py:
```diff
@@ -1,35 +1,135 @@
 __all__ = ['ncSilehubbard']
 
 
+def get_sile(file, *args, **kwargs):
```
Hi Nick, I'm having some problems when trying to implement the lines from above (mainly using `sisl` to add `ncsilehubbard` as a `sile` and then using the functions from `ncsilehubbard` to read the densities, U, etc).

I'm probably doing something wrong but I have found a problem when trying to read a file whose path is in another folder (when the file is in the working directory it runs normally).

For instance, when I run this peace of code:

```python
import sisl
from hubbard import HubbardHamiltonian, ncsile, sp2
import numpy as np

# Build sisl Geometry object only for a subset of atoms
molecule = sisl.Geometry([0,0,0], atoms=sisl.Atom(6), sc=[1,1,1])
molecule.sc.set_nsc([1, 1, 1])

# Build HubbardHamiltonian object
Hsp2 = sp2(molecule)
H = HubbardHamiltonian(Hsp2, U=3.5)
# Generate simple density
H.n = np.ones((2, H.sites))

# Create another folder
import os
os.mkdir('another_folder')

file_name = 'another_folder/test.nc'
H.write_density(file_name)

# If the file is not in the working directory then it crashes
sisl.io.add_sile(file_name, ncsile.ncSilehubbard, gzip=False)
HU = sisl.get_sile(file_name)
print('n: ', HU.read_density())
print('U: ', HU.read_U())
print('kT: ', HU.read_kT())
```

I get the following output:
```
  File "test-hubbard-2.py", line 23, in <module>
    print('n: ', HU.read_density())
  File "/home/sofia/anaconda3/lib/python3.8/site-packages/sisl-0.12.0.dev0-py3.8-linux-x86_64.egg/sisl/io/sile.py", line 463, in __getattr__
    return getattr(self.fh, name)
  File "src/netCDF4/_netCDF4.pyx", line 2945, in netCDF4._netCDF4.Dataset.__getattr__
  File "src/netCDF4/_netCDF4.pyx", line 2887, in netCDF4._netCDF4.Dataset.getncattr
  File "src/netCDF4/_netCDF4.pyx", line 1468, in netCDF4._netCDF4._get_att
  File "src/netCDF4/_netCDF4.pyx", line 1950, in netCDF4._netCDF4._ensure_nc_success
AttributeError: NetCDF: Attribute not found
```

If now I change `folder_name='test.nc'` then it works... I have no idea why this happens.

zerothi replied:
Yeah this isn't the way the `add_sile` is supposed to work.

The idea is that the suffix of the file describes which handle it should read from. A generic `*.nc` file suffix will be understood as a Siesta-nc file. That is why I think you should add another suffix that describes the hubbard file-format. It could be `HU.nc`, or `U.nc` or what-ever?

Then `add_sile("HU.nc", ncSileHubbard)` will use the `ncSileHubbard` class to read content whenever `get_sile("*.HU.nc")` gets invoked? Does this make sense?

sofiasanz replied:
Oh I see, I misunderstood the idea. Thanks!

I think it's fine to have that option. If the user saves the file with this suffix then it works (see commit d1a087f ) :)

zerothi commented on hubbard/ncsile.py:
```diff
@@ -1,35 +1,135 @@
         self._crt_dim(self, 'norb', n.shape[1])
 
         # Write variable n
         v = self._crt_var(g, 'n', ('f8', 'f8'), ('ncomp', 'norb'))
```
Yeah, I don't really care, it may just be mis-understood if you don't describe it. So an attribute would be nice ;)

zerothi commented on hubbard/hamiltonian.py:
```diff
@@ -57,6 +51,16 @@ def __init__(self, TBHam, n=0, U=0.0, q=(0., 0.), nkpt=[1, 1, 1], kT=1e-5):
+            s = H0.H.tocsr(0).data.tostring() + H0.H.tocsr(1).data.tostring()
+        else:
+            s = H0.H.tocsr(0).data.tostring()
+        self._hash_base = s
```
Ok. But the TB Hamiltonian changes on-site terms during the SCF. So effectively it changes while running... No?

sofiasanz replied:
No, the TB and the MFH Hamiltonians are saved separately so the TB Hamiltonian (even its onsite energies) doesn't change :)

We have these lines in the `HubbardHamiltonian` class:
```python
self.TBHam = TBHam
self.H = TBHam.copy()
```
Where `self.H` would save the MFH elements (so it includes the changes in the onsite energies) while `self.TBHam` remains untouched.

zerothi replied:
Ok! Thanks for making me understand ;)

zerothi commented on hubbard/ncsile.py:
```diff
@@ -11,25 +11,132 @@ class ncSilehubbard(sisl.SileCDF):
+        v1.info = 'Spin densities'
+        v2.info = 'Coulomb repulsion parameter in ' + units
+        v3.info = 'Temperature of the system in '+ units
         g.variables['n'][:] = n
```
Here just use `v1[:] = n` it is more clear I think. :)

sofiasanz commented:
Dears @zerothi and @tfrederiksen, following your ideas I made the following changes in the last commits (fa2ca6d to 6059f23):

1. Avoid using masked arrays (following comment https://github.com/dipc-cc/hubbard/pull/82#discussion_r738173952)
2. Added a specific test for the io functionalities (as pointed out in comment https://github.com/dipc-cc/hubbard/pull/82#pullrequestreview-790879448)
3. Added info when writing variables in the file and added units used in the problem (comment https://github.com/dipc-cc/hubbard/pull/82#discussion_r738176991)
4. Properly use `sisl` to read file using `get_sile` (as commented in https://github.com/dipc-cc/hubbard/pull/82#discussion_r738173310). 
5. Other minor changes such as renaming the function `__hash__` to `get_hash` in the `HubbardHamiltonian` class and `ncomp` to `nspin` (name of the dimension when writing the densities), etc.

zerothi commented:
Just for completeness sake, I would name your file `ncSileHubbard` (note upper-case Hubbard).

sofiasanz commented:
If @tfrederiksen agrees, I'll merge this PR then :)

Thank you both very much for your useful reviews!!

tfrederiksen commented on tests/test-hubbard-io.py:
```diff
@@ -0,0 +1,49 @@
+print('U: ', fh.read_U(group=None))
+print('kT: ', fh.read_kT(group=None))
+
+print('\n')
```
Here I would suggest to add another way to interact with the file handle:
```python
for g in fh.groups:
    print(g, fh.read_U(group=g))
```

sofiasanz replied:
Thanks! Added in commit 57f3a92

tfrederiksen reviewed (approved):
Looks great! Sorry for the delay.

sofiasanz commented:
I will merge this branch into the master branch then :)
